Test API updates

grybmadsci opened the pull request:
Another day, another `station_api` update... The primary focus here is mostly internal refactoring:
- `*Info`/`*Data` renamed to `*Descriptor` (`TestDescriptor` and `PhaseDescriptor`).  These contain static information about the respective thing, and do not change between executions.
- `*State` cleaned up extensively, particularly `PhaseData` -> `PhaseState` + `TestApi`
- `TestApi` established as the canonical object that is passed into test phases' first argument.  It's defined in `openhtf/__init__.py` and is thoroughly documented.  It also does not contain various internal guts like `PhaseData` used to.
- `Measurement` values are now merged into the `measurements list` (no more separate  `measured_values`).
- Monitors (not too many users) were updated a bit to have more consistent polling intervals.
- Lots of little nits updated.

There are still a handful of TODOs, and even more updates left to go before it's completely ready, but it's in a reasonable state right now, all tests pass, examples work, yay.  I'll hold off on merging this into `google/station_api` until tomorrow's update, after that you'll have to make comments elsewhere.

jettisonjoe commented on examples/example_plug.py:
```diff
@@ -80,4 +80,4 @@ def TearDown(self):
   def Increment(self):
     """Increment our value, return the new value."""
```
Docstring no longer accurate, it looks like.

grybmadsci replied:
done

jettisonjoe commented on openhtf/__init__.py:
```diff
@@ -367,20 +371,78 @@ def WithArgs(self, **kwargs):
     return self.func(**kwargs)
 
 
+class TestApi(collections.namedtuple('TestApi', [
```
Python style for abbrevs _might_ dictate `TestAPI` here instead of `TestApi`.

grybmadsci replied:
hm, can you cite a source?  i couldn't find anything in either direction for python, but google c++ style dictates:
`Prefer to capitalize acronyms as single words (i.e. StartRpc(), not StartRPC()).`

jettisonjoe commented on openhtf/__init__.py:
```diff
@@ -367,20 +371,78 @@ def WithArgs(self, **kwargs):
+        attribute should not be modified directly, use TestApi.Attach() or
+        TestApi.AttachFromFile() instead.
+
+    dut_id: This attribute provides getter and setter access to the DUT ID
```
Finally `dut_id` is a first-class citizen!

jettisonjoe commented on openhtf/exe/__init__.py:
```diff
@@ -51,9 +51,9 @@ def __init__(self, test_data, test_start, teardown_function=None):
-    self._plug_manager = plugs.PlugManager()
+        teardown_function
+        and openhtf.PhaseDescriptor.WrapOrCopy(teardown_function))
+    self._test_desc = test_data
```
Why not rename the arg, too?

grybmadsci replied:
ah, meant to, thanks.  also renamed to `test_descriptor` because unnecessary abbreviations are annoying.

jettisonjoe commented on openhtf/exe/test_state.py:
```diff
@@ -13,11 +13,18 @@
 
 
-"""TestState module for handling the lifetime of a test.
+"""Module for handling transient state of a running test.
```
Nice, much more clear/focused than the prev iteration.

fahhem commented on openhtf/__init__.py:
```diff
@@ -213,13 +239,13 @@ def Execute(self, test_start=None, loop=None):
+          self._test_desc, test_start, self._test_options.teardown_function)
+      _LOG.info('Executing test: %s', self.descriptor.code_info.name)
       self._executor.Start()
 
```
At some point you moved output callbacks being called out of this class, but I think it should remain here. They're attached to the test and it makes sense that `Test` call them. If `executor.Finalize()` returned the final `test_state`, then this would call the callbacks and add to history. There's a log in there that mentions history, but it's the `Test` that adds the `history.append_record` callback in there, so it's really twisted as-is.

grybmadsci replied:
good call, moved output callback looping back into here.

fahhem commented on openhtf/__init__.py:
```diff
@@ -334,19 +364,19 @@ class PhaseInfo(mutablerecords.Record(
 
   @classmethod
   def WrapOrCopy(cls, func):
```
now that `PhaseDescriptor.WrapOrCopy` is so long, and it gets used by test authors (to call `WithArgs` on a phase that doesn't use plugs, for instance), could we get a better-named function that calls out to this? Maybe `AsPhase`?

grybmadsci replied:
hm, i'll defer that to a separate change, could be as simple as an alias in `names.py`

grybmadsci commented:
going to merge this as it's in a mostly-working state.  i think i've addressed comments from this (and the previous) PR, but again since this is still a side working branch, things are still open for conversation, just keeping the ball rolling.
